L1.md

```diff
@@ -0,0 +1,154 @@
+# C++ streaming coalescing API's
+
+* Author(s): ctiller
+* Approver: a11r
+* Status: Draft
+* Implemented in: C++
+* Last updated: 2017/01/13
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/GfxB7nuV9GY
+
+## Abstract
+
+Provide C++ API's to coalesce:
+- initial metadata and first streaming message
+- final streaming message and trailing metadata
+into the same core batch.
+
+## Background
+
+Currently when making a streaming call using gRPC and sending a single message,
+three core batches are created: one for initial metadata, one for the message
+itself, and one for trailing metadata.
+
+Since the underlying transport has no knowledge that future messages are coming,
+it's forced to assume it should initiate writes immediately for each of them.
+In typical cases, this results in three sendmsg syscalls per stream.
+
+The current API allows some coalescing via WriteOptions. One can do:
+```c++
+auto s = stub->Foo(...);
+s->Write(first_message, WriteOptions().set_buffer_hint());
+s->WritesDone();
+```
+This is difficult to discover, and further more it performs two round-trips
+through the C Core stack, where it would be possible to do one.
+
+### Related Proposals:
+N/A
+
+## Proposal
+
+### Expand WriteOptions to include an end-of-stream bit
+```c++
+class WriteOptions {
+ public:
+  // ... existing interface ...
+
+  // last-message bit: indicates this is the last message in a stream
+  // client-side:  makes Write the equivalent of performing Write, WritesDone in
+  //               a single step
+  // server-side:  hold the Write until the service handler returns (sync api)
+  //               or until Finish is called (async api)
+  WriteOptions& set_last_message();
+  WriteOptions& clear_last_message();
+  bool is_last_message() const;
+};
+```
+
+### Add a convenience WriteLast method to ClientWriterInterface, ClientReaderWriterInterface, ServerWriterInterface, ServerReaderWriterInterface
+```c++
+// Perform Write, WritesDone in a single step
+void WriteLast(const W& msg, WriteOptions options) {
+  Write(msg, options.set_last_message());
+}
+```
+
+### Add a convenience WriteLast method to ClientAsyncWriterInterface, ClientAsyncReaderWriterInterface, ServerAsyncWriterInterface, ServerAsyncReaderWriterInterface
+```c++
+// Perform Write, WritesDone in a single step
+void WriteLast(const W& msg, WriteOptions options, void* tag) {
+  Write(msg, options.set_last_message(), tag);
+}
+```
+
+This will require finally exposing WriteOptions to async code also:
+```c++
+void Write(const W& msg, WriteOptions options, void* tag);
+```
+
+### Add a WriteAndFinish method to ServerAsyncWriterInterface, ServerAsyncReaderWriterInterface
+```c++
+// Perform Write, Finish in a single step
+void WriteAndFinish(const W& msg, WriteOptions options, const Status& status void* tag);
+```
+
+### Expand client streaming stub signatures to add an overload that takes the first message
```

Comment:
Per previous discussion, I would recommend adding an option either to the context or the initial metadata send that signals the same buffer-hint that we have in the write options.